File: moderate/card_number_validation.py
#!/usr/bin/env python2
# CodeEval - Card Number Validation
# https://www.codeeval.com/open_challenges/172/

from sys import argv

def check_number(number_str):
    number_str = number_str[::-1]
    sum_ = 0
    i = 0
    while i < len(number_str):
        if i % 2:
            prod = int(number_str[i]) * 2
            if prod > 9:
                sum_of_digits = 0
                prod_str = str(prod)
                for char in prod_str:
                    sum_of_digits += int(char)
                sum_ += sum_of_digits
            else:
                sum_ += prod
        else:
            sum_ += int( number_str[i] )

        i += 1
    
    if sum_ % 10:
        return 0
    else:
        return 1


def main():
    fh = open(argv[1], "rt")
    for line in fh:
        number_str = line.strip()
        number_str = number_str.replace(" ", "")
        print(check_number(number_str))

    fh.close()


if __name__ == "__main__":
    main()

File: moderate/card_number_validation.pl
#!/usr/bin/env perl
# CodeEval - Card Number Validation
# https://www.codeeval.com/open_challenges/172/

use strict;

sub check_number {
    my $number_str = shift @_;
    $number_str = reverse $number_str;
    my @number_str_arr = split //, $number_str;
    my $arr_length = @number_str_arr;
    my $sum = 0;
    my $i = 0;
    while ($i < $arr_length){
        if ($i %2){
            my $prod = $number_str_arr[$i] * 2;
            if ($prod > 9){
                my $sum_of_digits = 0;
                my @prod_arr = split //, $prod;
                foreach my $digit (@prod_arr){
                    $sum_of_digits += $digit;
                }
                $sum += $sum_of_digits;
            }
            else {
                $sum += $number_str_arr[$i] * 2;
            }
        }
        else {
            $sum += $number_str_arr[$i];
        }
        $i++;
    }

    if ($sum % 10){
        return 0;
    }
    else {
        return 1;
    }
}

open(INFILE, "<", $ARGV[0]) or die("Cannot open file $ARGV[0] for reading: $!");
while(my $number = <INFILE>) {
    next if($number =~ m/^\s$/);  
    $number =~ s/\s//gc;
    print check_number($number), "\n";
}
